lib/saluki-components/src/sources/dogstatsd/mod.rs

```diff
@@ -392,108 +429,181 @@ async fn process_listener(source_context: SourceContext, listener_context: Liste
-                // on each metric, if they came over UDS. This would then be utilized downstream in the pipeline by
-                // origin enrichment, if present.
-                if origin_detection {
-                    if let ConnectionAddress::ProcessLike(Some(creds)) = &peer_addr {
-                        for event in &mut event_buffer {
-                            if let Some(metric) = event.try_as_metric_mut() {
-                                metric
-                                    .metadata_mut()
-                                    .origin_entity_mut()
-                                    .set_process_id(creds.pid as u32);
-                            }
-                        }
-                    }
-                }
-
-                // Extract eventd events only if at least one is present in the event buffer.
-                let maybe_eventd_event_buffer = match event_buffer.has_data_type(DataType::EventD) {
-                    true => {
-                        let mut eventd_event_buffer = source_context.event_buffer_pool().acquire().await;
-                        eventd_event_buffer.extend(event_buffer.extract(Event::is_eventd));
-                        Some(eventd_event_buffer)
-                    }
-                    false => None,
-                };
-
-                // Extract service check events only if at least one is present in the event buffer.
-                let maybe_service_checks_event_buffer = match event_buffer.has_data_type(DataType::ServiceCheck) {
-                    true => {
-                        let mut service_check_event_buffer = source_context.event_buffer_pool().acquire().await;
-                        service_check_event_buffer.extend(event_buffer.extract(Event::is_service_check));
-                        Some(service_check_event_buffer)
-                    }
-                    false => None,
-                };
+        let mut buffer = io_buffer_pool.acquire().await;
+        debug!(capacity = buffer.remaining_mut(), "Acquired buffer for decoding.");
 
-                trace!(%listen_addr, %peer_addr, events_len = n, "Forwarding events.");
+        // If our buffer is full, we can't do any reads.
+        if !buffer.has_remaining_mut() {
+            // return Err(DeserializerError::BufferFull {
+            //     remaining: buffer.remaining(),
+            // });
+            // try to get a new buffer on the next iteration?
+            continue;
+        }
 
-                if let Err(e) = source_context.forwarder().forward_named("metrics", event_buffer).await {
-                    error!(%listen_addr, %peer_addr, error = %e, "Failed to forward metric events.");
-                }
+        // Try filling our buffer from the underlying reader first.
+        debug!("About to receive data from the stream.");
+        let (bytes_read, peer_addr) = stream.receive(&mut buffer).await.expect("TODO");
```

Comment:
Can we just flesh this back out to emit an error log and then break out of the receive loop? (i.e. lift up the code from the previous lines 491-493)